lib/widgets/compose_box.dart

```diff
@@ -1069,103 +1103,127 @@ class _ComposeBoxLayout extends StatelessWidget {
+      _AttachFileButton(contentController: content, contentFocusNode: contentFocusNode),
+      _AttachMediaButton(contentController: content, contentFocusNode: contentFocusNode),
+      _AttachFromCameraButton(contentController: content, contentFocusNode: contentFocusNode),
     ];
 
-    return _ComposeBoxContainer(
-      child: Column(children: [
-        Padding(
-          padding: const EdgeInsets.symmetric(horizontal: 8),
-          child: Theme(
-            data: inputThemeData,
-            child: Column(children: [
-              if (topicInput != null) topicInput!,
-              contentInput,
+    final topicInput = this.topicInput();
+    return Column(children: [
+      Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 8),
+        child: Theme(
+          data: inputThemeData,
+          child: Column(children: [
+            if (topicInput != null) topicInput,
+            contentInput(),
+          ]))),
+      SizedBox(
+        height: _composeButtonSize,
+        child: IconButtonTheme(
+          data: iconButtonThemeData,
+          child: Row(
+            mainAxisAlignment: MainAxisAlignment.spaceBetween,
+            children: [
+              Row(children: composeButtons),
+              sendButton(),
             ]))),
-        SizedBox(
-          height: _composeButtonSize,
-          child: IconButtonTheme(
-            data: iconButtonThemeData,
-            child: Row(
-              mainAxisAlignment: MainAxisAlignment.spaceBetween,
-              children: [
-                Row(children: composeButtons),
-                sendButton,
-              ]))),
-      ]));
+    ]);
   }
 }
 
-abstract class ComposeBoxController<T extends StatefulWidget> extends State<T> {
-  ComposeTopicController? get topicController;
-  ComposeContentController get contentController;
-  FocusNode get contentFocusNode;
-}
-
 /// A compose box for use in a channel narrow.
 ///
 /// This offers a text input for the topic to send to,
 /// in addition to a text input for the message content.
-class _StreamComposeBox extends StatefulWidget {
-  const _StreamComposeBox({super.key, required this.narrow});
+class _StreamComposeBoxBody extends _ComposeBoxBody {
+  _StreamComposeBoxBody({required this.narrow, required this.controller});
 
-  /// The narrow on view in the message list.
+  @override
   final ChannelNarrow narrow;
 
   @override
-  State<_StreamComposeBox> createState() => _StreamComposeBoxState();
+  final StreamComposeBoxController controller;
+
+  @override Widget topicInput() => _TopicInput(
+    streamId: narrow.streamId,
+    controller: controller.topic,
+    focusNode: controller.topicFocusNode,
+    contentFocusNode: controller.contentFocusNode,
+  );
+
+  @override Widget contentInput() => _StreamContentInput(
+    narrow: narrow,
+    topicController: controller.topic,
+    controller: controller.content,
+    focusNode: controller.contentFocusNode,
```

Comment:
A further step that I think could be helpful is if ComposeBox grew an InheritedWidget and a static `of` method. Then instead of passing these controller pieces down through each layer, the descendant widgets that need them could say `ComposeBox.of(context).controller` and get what they need.

Or maybe the `of` method would even just return the controller ­— that's the meaningful part of the state anyway.

Reply:
Alternatively, the widgets could continue taking this information as parameters — but as a single parameter that is the whole controller. That seems like it'd also clean this up a lot.

Either way, one thing I'm thinking of in particular is how #924 adds `enabled` and `sendMessageError`, and has to pass them down through layers in a bunch of places (especially `enabled`). So ideally I'd like to see a refactor that makes it so a change like #924 can add a field to the controller and then not have to add references to it at a bunch of widget constructors but still be able to use it at various widgets deeper inside the compose box, like the various buttons.

Reply:
With any of these approaches, we still need to listen to the possible updates. Our current setup for listener has a fine granularity — `_ContentInputState` listens to `ComposeContentController` and a `FocusNode`, `_StreamContentInputState` just listens to `ComposeTopicController`…

Continuing this trend, we may have individual listeners for `enabled` and `sendMessageError` as well, while turning `_TopicInput` and probably some others into `StatefulWidget`s.

With `InheritedWidget`, maybe we can use `BuildContext.dependOnInheritedWidgetOfExactType` to simplify that part.

Reply:
I think the existing ways we have things listen don't need to change.

The one thing worth avoiding with fine vs. coarse granularity here is that it'd be good to minimize how much needs to rebuild every time the user types another character. So things that don't need to depend on the text controllers should avoid doing so. But if things depend on `enabled` or `sendMessageError` when they don't strictly need to, that isn't a problem - those change much less frequently.

Reply:
> InheritedWidget

I almost made a commit like this, but then noticed the branch was getting pretty long. :) I also noticed the `InheritedWidget` API prompting me to think about correctly handling changes to the value (the controller), such as not calling `dependOnInheritedWidgetOfExactType` in an `initState` method. It seemed like something to consider carefully, but I didn't have a clear idea of when the controller would change (or if/when we'd want it to), so I didn't go ahead with it.